src/kvstore/HBaseEngine.cpp

```diff
@@ -0,0 +1,353 @@
+    if (code == ResultCode::ERR_IO_ERROR) {
+        LOG(ERROR) << "Range Failed: the HBase I/O error.";
+    }
+    return code;
+}
+
+
+ResultCode HBaseEngine::prefix(const std::string& prefix,
+                               std::unique_ptr<KVIterator>* storageIter) {
+    std::string startRowKey, endRowKey;
+    startRowKey.reserve(kMaxRowKeyLength);
+    endRowKey.reserve(kMaxRowKeyLength);
+    startRowKey.append(prefix);
+    endRowKey.append(prefix);
+    for (size_t n = 0; n < kMaxRowKeyLength - prefix.size(); n++) {
+        startRowKey.append(reinterpret_cast<const char*>(&kFillMin), sizeof(uint8_t));
+        endRowKey.append(reinterpret_cast<const char*>(&kFillMax), sizeof(uint8_t));
+    }
+    ResultCode code = this->range(startRowKey, endRowKey, storageIter);
+    if (code == ResultCode::ERR_IO_ERROR) {
+        LOG(ERROR) << "Prefix Failed: the HBase I/O error.";
+    }
+    return code;
+}
+
+
+ResultCode HBaseEngine::remove(const std::string& key) {
+    auto rowKey = this->getRowKey(key);
+    ResultCode code = client_->remove(tableName_, rowKey);
+    if (code == ResultCode::ERR_IO_ERROR) {
+        LOG(ERROR) << "Remove Failed: " << key << ", the HBase I/O error.";
+    }
+    return code;
+}
+
+
+ResultCode HBaseEngine::multiRemove(std::vector<std::string> keys) {
+    std::vector<std::string> rowKeys;
+    for (size_t i = 0; i < keys.size(); i++) {
+        auto rowKey = this->getRowKey(keys[i]);
+        rowKeys.emplace_back(rowKey);
+    }
+    ResultCode code = client_->multiRemove(tableName_, rowKeys);
+    if (code == ResultCode::ERR_IO_ERROR) {
+        LOG(ERROR) << "MultiRemove Failed: the HBase I/O error.";
+    }
+    return code;
+}
+
+
+ResultCode HBaseEngine::removeRange(const std::string& start,
+                                    const std::string& end) {
+    std::unique_ptr<kvstore::KVIterator> iter;
+    ResultCode rangeCode = this->range(start, end, &iter);
+    if (rangeCode == ResultCode::SUCCEEDED) {
+        std::vector<std::string> keys;
+        while (iter->valid()) {
+            keys.emplace_back(iter->key());
+            iter->next();
+        }
+        ResultCode code = this->multiRemove(keys);
+        if (code == ResultCode::ERR_IO_ERROR) {
+            LOG(ERROR) << "RemoveRange Failed: the HBase I/O error.";
+        }
+        return code;
+    } else if (rangeCode == ResultCode::ERR_IO_ERROR) {
+        LOG(ERROR) << "RemoveRange Failed on Range: the HBase I/O error.";
+    }
+    return rangeCode;
+}
+
+
+int32_t HBaseEngine::totalPartsNum() {
+    return partsNum_;
+}
+
+
+void HBaseEngine::addPart(PartitionID partId) {
+    UNUSED(partId);
+    LOG(INFO) << "HBaseEngine does not support the \"addPart\" method.";
+}
+
+
+void HBaseEngine::removePart(PartitionID partId) {
+    UNUSED(partId);
+    LOG(INFO) << "HBaseEngine does not support the \"removePart\" method.";
+}
+
+
+std::vector<PartitionID> HBaseEngine::allParts() {
+    std::vector<PartitionID> parts;
```

Comment:
Here is no `Partition` in `HBase` and don't support `addPart` and `removePart`.
Why return `0` as result ?

Reply:
There is a **ONE** big partition.   FYI. https://github.com/vesoft-inc/nebula/issues/179#issuecomment-482502357
Another reason is  just for overriding interface of KVEngine.